### ISSUE TITLE
iwyu suggests forward declaration when using typeid

### ISSUE THREAD

**Issue by cottonxray** [include-what-you-use/include-what-you-use#1168]
Thanks for all your work on iwyu. I can't imagine sorting out my dependency messes without it :)

I've recently seen an issue where iwyu will suggest a forward-declaration instead of a `#include` for `Dependency` here:

```c++
// user/user.cpp
#include <dependency.h>
#include <typeindex>

auto useTypeIndexOfDependency(const Dependency &dependency) -> void {
  std::type_index type = typeid(dependency);
}
```

```
$ iwyu_tool -p build user/user.cpp

/home/matthew/Documents/iwyu/minimal-example/user/user.cpp should add these lines:
class Dependency;

/home/matthew/Documents/iwyu/minimal-example/user/user.cpp should remove these lines:
- #include <dependency.h>  // lines 1-1

The full include-list for /home/matthew/Documents/iwyu/minimal-example/user/user.cpp:
#include <typeindex>  // for type_index
class Dependency;
---
```
If I follow its suggestion, I get 
```
[ 50%] Building CXX object user/CMakeFiles/user.dir/user.cpp.o
/home/matthew/Documents/iwyu/minimal-example/user/user.cpp: In function ‘void useTypeIndexOfDependency(const Dependency&)’:
/home/matthew/Documents/iwyu/minimal-example/user/user.cpp:5:33: error: invalid use of incomplete type ‘class Dependency’
    5 |   std::type_index type = typeid(dependency);
      |                                 ^~~~~~~~~~
/home/matthew/Documents/iwyu/minimal-example/user/user.cpp:2:7: note: forward declaration of ‘class Dependency’
    2 | class Dependency;
      |       ^~~~~~~~~~
make[2]: *** [user/CMakeFiles/user.dir/build.make:82: user/CMakeFiles/user.dir/user.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:134: user/CMakeFiles/user.dir/all] Error 2
make: *** [Makefile:103: all] Error 2
```
I believe this is because you cannot apply `typeid` to an incomplete type.

I tested this with
```
$ include-what-you-use --version
include-what-you-use 0.15 based on Debian clang version 11.0.1-2
```
which is the current version on Debian Stable.

I've uploaded a full minimal example [here](https://github.com/cottonxray/iwyu-typeid-potential-issue).

I would be happy to spend some time to contribute a fix for this, if that would help. Equally, if there's any extra information that I can provide, then let me know.

But I wanted to check:
* do you also think it's an issue, or have I missed something?
* if you'd like me to work on a fix - where would be the best place to make this change in the code?

Thanks again for all your work on this tool.

**Comment by kimgr** [comment]
Thanks for the report! I think there's just a visitor missing for the `typeid` -- adding a `VisitCXXTypeidExpr` [1] and explicitly reporting the type used should just work :tm:.

[1] https://clang.llvm.org/doxygen/classclang_1_1CXXTypeidExpr.html

**Comment by cottonxray** [comment]
Thanks, that saved me a lot of digging around the code!
I've written up a test and fix locally. It seems to work, although I'm not sure if I've implemented the visitor in the best way.
I have to get approval from my employer before I can contribute any code, so once they reply I'll either raise the PR or post back here.
I've submitted the request already, so hopefully it won't take long to hear back from them.
